File: rust/src/day1.rs
use std::{collections::HashMap, iter::Iterator};

pub fn day1_part1(lines: impl Iterator<Item = String>) -> i64 {
    let (mut left, mut right) = parse_input(lines);
    left.sort();
    right.sort();
    let mut res = 0;
    for i in 0..left.len() {
        let dist = left[i] - right[i];
        res += dist.abs();
    }
    res
}

pub fn day1_part2(lines: impl Iterator<Item = String>) -> i64 {
    let (left, right) = parse_input(lines);
    let mut m = HashMap::new();
    for r in right.iter() {
        m.entry(r).and_modify(|e| *e += 1).or_insert(1);
    }
    let mut res = 0;
    for l in left.iter() {
        res += l * m.get(l).unwrap_or(&0);
    }
    res
}

fn parse_input(input: impl Iterator<Item = String>) -> (Vec<i64>, Vec<i64>) {
    let mut left = Vec::new();
    let mut right = Vec::new();
    for line in input {
        let mut split = line.split_whitespace();
        let (l, r) = (
            split.next().unwrap().parse::<i64>().unwrap(),
            split.next().unwrap().parse::<i64>().unwrap(),
        );
        left.push(l);
        right.push(r);
    }
    (left, right)
}

File: go/day1.go
package main

import (
	"iter"
	"slices"
)

func day1Part1(lines iter.Seq[string]) int {
	left, right := parseInput(lines)

	slices.Sort(left)
	slices.Sort(right)
	res := 0
	for i := 0; i < len(left); i++ {
		res += abs(left[i] - right[i])
	}

	return res
}

func day1Part2(lines iter.Seq[string]) int {
	left, right := parseInput(lines)

	m := make(map[int]int, len(left))
	for _, nr := range right {
		m[nr] += 1
	}
	res := 0
	for _, nl := range left {
		res += nl * m[nl]
	}

	return res
}

func parseInput(input iter.Seq[string]) ([]int, []int) {
	var left, right []int
	for l := range input {
		nums := parseNums([]byte(l))
		left = append(left, nums[0])
		right = append(right, nums[1])
	}
	return left, right
}
